File: _bin/jp-time.js
#!/usr/bin/env node

const IMAWA = '今は'
const REIJI = '零時'
const SHOUGO = '正午'
const GOZEN = '午前'
const GOGO = '午後'
const JI = '時'
const FUN = '分'
const HAN = '半'
const CHOUDO = 'ちょうど'
const GORO = 'ごろ'
const DESU = 'です。'

function main() {
  const phrase = generatePhrase(new Date())
  if (!process.stdout.isTTY) console.error(phrase)
  console.log(phrase)
}

function generatePhrase(date) {
  let hh = date.getHours()
  let mm = date.getMinutes()

  const imawa = coinFlip() ? IMAWA : ''

  const choudo = (mm === 0 || mm === 30) && coinFlip() ? CHOUDO : ''

  let goro = ''
  const rounded = coinFlip() && round(hh, mm)
  if (rounded && (rounded.hh !== hh || rounded.mm !== mm)) {
    ;({ hh, mm } = rounded)
    goro = GORO
  }

  let ampm = ''
  if (hh % 12 > 0 && coinFlip()) {
    ampm = hh < 12 ? GOZEN : GOGO
    hh %= 12
  }

  const hour
    = hh === 0 && mm === 0 ? REIJI
    : hh === 12 && mm === 0 ? SHOUGO
    : `${hh}${JI}`

  const minute
    = mm === 0 ? ''
    : mm === 30 && (goro || coinFlip()) ? HAN
    : `${mm}${FUN}`

  const hhmm = `${hour}${minute}`
  const time = ampm || coinFlip() ? `${hhmm}${choudo}` : `${choudo}${hhmm}`

  return `${imawa}${ampm}${time}${goro}${DESU}`
}

function round(hh, mm) {
  if (mm < 5) {
    mm = 0
  } else if ((25 < mm && mm < 30) || (30 < mm && mm < 35)) {
    mm = 30
  } else if (55 < mm) {
    hh = (hh + 1) % 24
    mm = 0
  }

  return { hh, mm }
}

function coinFlip() {
  return Math.random() > 0.5
}

main()

File: _bin/jp-time.py
#!/usr/bin/env python3

from datetime import datetime
import random
import sys


IMAWA = '今は'
REIJI = '零時'
SHOUGO = '正午'
GOZEN = '午前'
GOGO = '午後'
JI = '時'
FUN = '分'
HAN = '半'
CHOUDO = 'ちょうど'
GORO = 'ごろ'
DESU = 'です。'


def main():
    phrase = generate_phrase(datetime.now())
    if not sys.stdout.isatty(): print(phrase, file=sys.stderr)
    print(phrase)


def generate_phrase(datetime):
    hour = datetime.hour
    minute = datetime.minute

    imawa = IMAWA if coinflip() else ''

    choudo = CHOUDO if (minute == 0 or minute == 30) and coinflip() else ''

    goro = ''
    rounded = coinflip() and round(hour, minute)
    if rounded and rounded != (hour, minute):
        hour, minute = rounded
        goro = GORO

    ampm = ''
    if (hour % 12) > 0 and coinflip():
        ampm = GOZEN if hour < 12 else GOGO
        hour %= 12

    hh = (REIJI if (hour, minute) == (0, 0)
        else SHOUGO if (hour, minute) == (12, 0)
        else f'{hour}{JI}')

    mm = ('' if minute == 0
        else HAN if minute == 30 and (goro or coinflip())
        else f'{minute}{FUN}')

    hhmm = f'{hh}{mm}'
    time = f'{hhmm}{choudo}' if ampm or coinflip() else f'{choudo}{hhmm}'

    return f'{imawa}{ampm}{time}{goro}{DESU}'


def round(hour, minute):
    if minute < 5:
        minute = 0
    elif 25 < minute < 30 or 30 < minute < 35:
        minute = 30
    elif 55 < minute:
        hour = (hour + 1) % 24
        minute = 0

    return hour, minute


def coinflip():
    return random.random() > 0.5


if __name__ == '__main__':
    main()
